File: greedy/02_Jump_Game/go/02_Jump_Game.go
package main

import "fmt"

func canJump(nums []int) bool {
	reachable := 0
	for i := 0; i < len(nums); i++ {
		if i > reachable {
			return false
		}
		reachable = max(reachable, i+nums[i])
	}
	return reachable >= len(nums)-1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	testCases := []struct {
		nums   []int
		expect bool
	}{
		{[]int{2, 3, 1, 1, 4}, true},
		{[]int{3, 2, 1, 0, 4}, false},
		{[]int{1, 3, 4, 1, 1, 2, 1}, true},
		{[]int{1, 3, 4, 2, 1, 1, 0, 1, 1}, false},
		{[]int{0}, true},
		{[]int{2, 0, 0}, true},
		{[]int{1, 0, 1, 0}, false}, // New test case
		{[]int{1, 1, 1, 1}, true},
		{[]int{1, 1, 0, 1}, false},
		{[]int{5, 0, 0, 0, 0, 0, 1}, false},
	}

	for i, tc := range testCases {
		fmt.Printf("Test Case %d: nums=%v\n", i+1, tc.nums)
		res := canJump(tc.nums)
		fmt.Printf("Got: %t\n", res)
		fmt.Printf("Expected: %t\n", tc.expect)
		if res != tc.expect {
			fmt.Println("FAIL")
		} else {
			fmt.Println("PASS")
		}
		fmt.Println()
	}
}

File: greedy/02_Jump_Game/02_Jump_Game.js
function canJump(nums) {
  let reachable = 0;
  for (let i = 0; i < nums.length; i++) {
    if (i > reachable) {
      return false;
    }
    reachable = Math.max(reachable, i + nums[i]);
  }
  return reachable >= nums.length - 1;
}

const testCases = [
  { nums: [2, 3, 1, 1, 4], expect: true },
  { nums: [3, 2, 1, 0, 4], expect: false },
  { nums: [1, 3, 4, 1, 1, 2, 1], expect: true },
  { nums: [1, 3, 4, 2, 1, 1, 0, 1, 1], expect: false },
  { nums: [0], expect: true },
  { nums: [2, 0, 0], expect: true },
  { nums: [1, 0, 1, 0], expect: false },
  { nums: [1, 1, 1, 1], expect: true },
  { nums: [1, 1, 0, 1], expect: false },
  { nums: [5, 0, 0, 0, 0, 0, 1], expect: false },
  { nums: [2, 5, 0, 0], expect: true },
];

for (let i = 0; i < testCases.length; i++) {
  const tc = testCases[i];
  console.log(`Test Case ${i + 1}: nums=${JSON.stringify(tc.nums)}`);
  const res = canJump(tc.nums);
  console.log(`Got: ${res}`);
  console.log(`Expected: ${tc.expect}`);
  if (res !== tc.expect) {
    console.log("FAIL");
  } else {
    console.log("PASS");
  }
  console.log();
}
